### PR TITLE
Inline calculations

### LINKED ISSUE #57: Switch to in-place `criterion_and_gradient!`
As noted in #54, a switch to an in-place gradient calculation could speed up performance by avoiding array allocations:

> But if in the long run you consider switching to a unified in-place _criterion_and_gradient!(grad::Union{AbstractVector, Nothing}, ...)_, which would skip gradient calculation if _grad === nothing_, this PR would be irrelevant.
> Right now the profiling shows that for a 62x24 matrix rotation ~10% is spent in the array allocation (both gradient calculation and projection).

### PR DESCRIPTION
**Pull request by alyst**
Switches to in-line `criterion_and_gradient!()`, as well as in-line `gradient_f!()` and in-line `project_X!()`.

I have tried to reduce temporary array allocations by using methods like 5-arg `mul!()` or `axpy!()` and made some tweaks in specific rotation methods.
But more in-depth code analysis + profiling would be required to tune the performance (in future PRs).

Note that I have changed the reference Quartimax loadings (`pub` matrix) in the test: after the `crieterion_and_gradient!()` changes it started to fail, and increasing `atol` did not help.
So I have checked what is the output of the *GPArotation* package (from the same authors as the original 2005 publication), and it was also different from `pub`.
The Quartimax criterion based on the original `pub` loadings is smaller (by ~1e-6), but I suspect that is the result of the rounding error, and the corresponding rotation matrix would have been non-orthogonal.
The criterion difference between GPArotation and Quartimax was ~1e-16.

Fixes #57

### CODE REVIEW

**Comment by alyst** [comment]
This is how profiling looks for varimax of 732x30 matrix after this PR:
image

File: src/methods/methods.jl [hunk 1]
```diff
@@ -16,21 +16,27 @@ abstract type RotationMethod{T<:RotationType} end
...
-    return Q, ∇Q
+criterion_and_gradient!
+
+# fallback method that uses AutoDiff
+function criterion_and_gradient!(∇Q, method::RotationMethod, Λ::AbstractMatrix)
+    if !isnothing(∇Q)
+        gradient!(Reverse, ∇Q, Base.Fix1(criterion, method), Λ)
+    end
+    return criterion(method, Λ)
 end
```
Comment (p-gw) [inline src/methods/methods.jl:40]: When do we need to call `criterion_and_gradient!` with `∇Q = nothing`? In the rotate function the gradient is always supplied and a user would simply call `criterion` if gradients are not needed. Am I missing something?

Comment (alyst) [reply src/methods/methods.jl:40]: In this PR, `criterion_and_gradient!()` becomes the only function that has to be implemented for the rotation method, and `criterion()` is just a wrapper that calls `criterion_and_gradient!(nothing, method, L)`. That way we eliminate code duplication between `criterion()` and `criterion_and_gradient!()`; and since Julia will compile specialized version of `criterion_and_gradient!(nothing, method, L)`, it should effectively eliminate the codepaths related to gradient evaluation.

There's one exception, though: `ComponentLoss` only implements `criterion()`, and in that case the autodiff-based fallback `criterion_and_gradient!()` method would be called -- the one we are discussing now.

Comment (alyst) [reply src/methods/methods.jl:40]: One thing that I don't like with the current state of PR is that if a specific rotation method does not implement neither `criterion()` nor `criterion_and_gradient!()`, calling `criterion_and_gradient!()` would result in a deadlock, since the fallback implementation would call the `criterion()` wrapper, and the wrapper would call the fallback `criterion_and_gradient!()` and so on.

To resolve this, we can change the fallback `criterion_and_gradient!()` to call the internal `criterion_only()` method.
So if the rotation method (i.e. *ComponentLoss*) does not want to implement the full `criterion_and_gradient!()`, it would have to implement `criterion_only()`.

Comment (alyst) [reply src/methods/methods.jl]: Ok, I've updated the PR so that `criterion()` is always a wrapper for `criterion_and_gradient!(nothing, ...)`.
And those *RotationMethod* subtypes, which opt to only implement criterion calculation and rely on autodiff for gradient, would have to implement `criterion_only()` method.

Comment (p-gw) [reply src/methods/methods.jl]: I don't really see why can't keep the design as is. 
There are 3 possibilities of implementation: 

1. Implement `criterion` but not `criterion_and_gradient!`. In this case gradient calculation will be handled by the fallback method of `criterion_and_gradient!` using autodiff.
2. Implement `criterion_and_gradient!` but not `criterion`. Then the rotation just uses the implemented method
3. Implement both `criterion` and `criterion_and_gradient!`. Then rotation will use the implemented method and users have the additional convenience of calling `criterion`.

Comment (alyst) [reply src/methods/methods.jl]: Well, you see what have already happened with varimax: *criterion* and *criterion_and_gradient* ended up having different code.
*criterion_and_gradient!* solves code duplication problem, which would also simplify maintenance in the long run. You would not have to check in each PR that the contributor has properly modified the two versions.
Again, from the user perspective *criterion()* will just work as before (calling *criterion_and_gradient!()* under the hood), and there should be no performance penalty.
*criterion_only* is for those so far rare cases, where gradient calculation is done by autodiff. I added it for 2 reasons: a - to avoid deadlocks and provide proper error message when the rotation method does not implement anything, b - to separate internal RotationMethod calls from the public API, and allow user-facing *criterion* implement common logic in the future (e.g. input parameter checks).

Comment (alyst) [reply src/methods/methods.jl]: Actually, I've realized that `criterion_only()` is not needed after all.
For autodiff-based gradients one has to implement only `criterion_and_gradient!(nothing, ...)`, while `criterion_and_gradient!(grad::Vector, ...)` would be the fallback autodiff-based method that calls `criterion_and_gradient!(nothing, ...)` to calculate the gradient.
And in general, it should always be possible to split `criterion_and_gradient!` into 2 separate definitions for `criterion_and_gradient!(nothing, ...)` and `criterion_and_gradient!(grad::Vector, ...)` for a particular rotation method.
So `criterion_and_gradient!()` would be the only function that is needed to implement for a rotation method, and `criterion()` would be the convenience wrapper.

I can update the PR with that changes. I can also rollback and reenable `criterion()` overloading if that's your strong preference.
I just think it is a bit less convenient in the long run.

Comment (alyst) [reply src/methods/methods.jl]: I've replaced `criterion_only(...)` with `criterion_and_gradient!(nothing, ...)`

Comment (alyst) [reply src/methods/methods.jl]: @p-gw So with the last modifications the user can potentially overload the `criterion()` to calculate the criterion of a specific rotation method, but I'd recommend to keep `criterion()` as a wrapper and overload `criterion_and_gradient!()` instead, that's what is written in the updated docs. I have also further optimizations of criterion and gradient calculations for individual rotation methods, but as this PR is already quite large, I'd rather put them in a separate PR when/if this PR is merged.

**Comment by p-gw** [comment]
> @p-gw Gentle ping :)

I'll most likely get to review it this week.

File: test/methods.jl [hunk 2]
```diff
@@ -108,7 +109,7 @@ end
         @test_throws ArgumentError Concave(0)
         @test_throws ArgumentError Concave(-2)
 
-        method = Absolmin(0)
+        method = Absolmin(1e-5)
```
Comment (p-gw) [inline test/methods.jl:112]: Why do we need to change `epsilon` here?

File: test/methods.jl [hunk 3]
```diff
@@ -284,23 +285,36 @@ end
...
+        #    0.9021319 0.1038604
+        #    0.8765090 0.1712805
+        #    0.3155758 0.8764747
+        #    0.2511265 0.7734879
+        #    0.1980102 0.7146775
+        #    0.3078601 0.6593331
+        #]
+        # loadings obtained with GPArotation v2024.3
+        # quartimax(A, eps=1e-8, maxit=50000, randomStarts=10)
+        gpa = [
```
Comment (p-gw) [inline test/methods.jl:305]: Do you think it is a good idea to test against a battle tested package such as GPArotation for a given problem?

File: test/methods.jl [hunk 4]
```diff
@@ -329,4 +343,12 @@ end
             init,
         )
     end
+
+    @testset "Missing criterion implementation" begin
+        struct NoCriterion <: RotationMethod{Orthogonal} end
+
+        @test_throws "NoCriterion does not implement" criterion(NoCriterion(), randn(6, 6))
+        # Enzyme.jl would refuse to autodiff because it detects that fallback criterion() throws an error
+        @test_throws "Function to differentiate" criterion_and_gradient!(randn(6, 5), NoCriterion(), randn(6, 5))
```
Comment (p-gw) [inline test/methods.jl:352]: I don't really like that we are relying on Enzyme to catch the error in this edge case. However, I have no idea how to catch them with the current implementation of `criterion` and `criterion_and_gradient!`. Any ideas?

File: test/rotate.jl [hunk 5]
```diff
@@ -3,11 +3,11 @@
     @test_throws ArgumentError rotate(A, Varimax(), init = rand(10, 10))
     @test_throws ArgumentError rotate(A, Varimax(), init = rand(8, 2))
 
-    rot_default_init = rotate(A, Varimax())
-    rot_identity_init = rotate(A, Varimax(); init)
-    @test loadings(rot_default_init) ≈ loadings(rot_identity_init)
-    @test rotation(rot_default_init) ≈ rotation(rot_identity_init)
-    @test factor_correlation(rot_default_init) ≈ factor_correlation(rot_identity_init)
+    rot_default_init = rotate(A, Varimax(), atol=1e-7)
```
Comment (p-gw) [inline test/rotate.jl:6]: `atol` needs to be changed to `g_atol` after rebasing on the current version of main since #55 has been merged

**Review by p-gw — CHANGES_REQUESTED** [review]
The overall changes look good to me, nice speed up in `rotate` as well. Before we can merge this we need to rebase on the current main branch.

File: test/methods.jl [hunk 6]
```diff
@@ -1,27 +1,28 @@
...
-    @test ∇Q isa AbstractMatrix{<:Real}
-    @test size(∇Q) == size(Λ)
-
-    # test the criterion() method if available
-    if applicable(criterion, method, Λ)
-        Q2 = criterion(method, Λ)
-        @test Q2 isa Real
-        @test Q ≈ Q2
-    end
+    @test all(isfinite, ∇Q)
```
[the remark this answers is not in the document]
Comment (alyst) [reply test/methods.jl:5]: It cannot be guaranteed for any input, but should be true for most of the input points.
And as currently we are only testing a single point, I think we can guarantee that this point would give finite values.
I just wanted to have a bit stricter check for calculation errors, not just that it updates ∇Q in principle.

File: test/methods.jl (hunk 4 above)
Comment (alyst) [reply test/methods.jl:352]: We don't rely on Enzyme, it's just too smart and recognizes that the fallback implementation would throw an "not implemented" error, so it throws an error first. :)
So I don't see a problem in terms of the design.
We may need to update this particular test if the Enzyme behavior changes.
We can make it less specific already now, but I don't see it as a big problem.

File: test/methods.jl (hunk 3 above)
Comment (alyst) [reply test/methods.jl:305]: I think it is, it could also be done more consistently for all the methods that the packages have in common.

One general consideration here is if we are comparing the rotation results against some reference, which has limited precision (e.g. the results from 2005 paper) -- that limited precision implies that the corresponding rotation may slightly violate the orthogonality constraints, and the corresponding reference criterion might be accidentally more optimal because the "point" is outside of the domain.
So ideally all reference rotations should contain as many digits after 0 as possible.

**Comment by alyst** [comment]
> The overall changes look good to me, nice speed up in `rotate` as well. Before we can merge this we need to rebase on the current main branch.

Thank you for the review! I've rebased the PR, the CI tests pass except for nightly, which is due to Ezyme problems not related to the PR.

File: test/methods.jl (hunk 2 above)
Comment (p-gw) [reply test/methods.jl:112]: Then we should disallow any `epsilon <= 0` at construction time of the criterion.

**Review by p-gw — APPROVED** [review]
